### PR TITLE
`Development`: Improve LTI authentication

### PR DESCRIPTION
**Pull request by maximiliansoelch**
#### General

- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).

### Motivation and Context

Currently, users need to login to Artemis when they do a LTI launch from an external system even in the case they are already logged in. This issue is caused by the fact that the LTILaunchFilter is placed before the JWTFilter. As the authentication is checked within the JWTFilter, this information is just not yet present in the LTILaunchFilter, requiring a user login every time.

### Description

- Add LTI Filter after JWTFilter so authentication info is available at launch again
  image
- Remove code that sets a logout cookie for authenticated users
- Also allows to trust an external system (with the same user base) to allow for automatic login
- Adapts the tests to the changes

### Steps for Testing

Prerequisites:
- Access to Moodle
- Testserver 1

1. Make sure you are logged in to Artemis with a test user (artemis_test_user_{1-5})
2. Navigate to [Moodle](https://moodle.ase.in.tum.de) and login with the same test user (artemis_test_user_{1-5}, same PW as for the Artemis Test Server)
3. Go to My Courses and open the course TS1 - Artemis Feature Demo Course
4. Click on one of the exercise
5. Artemis should open in an iFrame and you should still be logged in with the same test user (In case Artemis opens in a separate browser window, this is ok, then the course setting is wrong, please focus on the aspect whether the login works automatically or not)

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Performance Review
- [ ] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [ ] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

## Summary by CodeRabbit

- **New Features**
	- Enhanced security architecture by integrating JWT-based authentication for LTI 1.3 handling.
	- Introduced a configuration option to trust external LTI systems, allowing for more flexible user authentication.

- **Bug Fixes**
	- Removed redundant checks related to user activation and LTI login requirements, simplifying session management.

- **Tests**
	- Updated test methods to reflect changes in LTI query parameter handling for new and existing users.

### CODE REVIEW

**Comment by MichaelOwenDyer** [comment]
I am going to steal the lock for TS1, hope that's okay :)

**Review by az108 — APPROVED** [review]
Tested on TS1 works as expected 👍

**Review by Jan-Thurner — CHANGES_REQUESTED** [review]
Tested on TS1. When logging in as TestUser2, and trying to access an exercise, I have to log in to artemis again manually. 
image

Also, the abundance of testCourses on moodle that are all named the same but with only one having the correct links makes the review process quite uncomfortable.

**Review by az108 — CHANGES_REQUESTED** [review]
does not work on TS1 for me aswell and trying to log in artemis trough the iframe does not work for me aswell

**Comment by raffifasaro** [comment]
> does not work on TS1 for me aswell and trying to log in artemis trough the iframe does not work for me aswell

I tested it on TS1 with 2 different devices and was logged in without problems, although Artemis is not displayed with an iFrame on the moodle site, but the user is directly forwarded to the Artemis page.

**Comment by Jan-Thurner** [comment]
> > does not work on TS1 for me aswell and trying to log in artemis trough the iframe does not work for me aswell
> 
> I tested it on TS1 with 2 different devices and was logged in without problems, although Artemis is not displayed within an iFrame on the moodle site, but the user is directly forwarded to the Artemis page.

This is what I meant with:
> Also, the multitude of testCourses on moodle that are all named the same but with only one having the correct links makes the review process quite uncomfortable.
The testing steps are quite confusing and the expected result does not occur no matter what course is used. Either the user is not logged in, or the course is not opened in an iFrame. I think that either something in the code still needs to be adjusted, or the testing steps need to be modified to match the expected outcome (if the way it is currently working is the intended way).

**Comment by raffifasaro** [comment]
> does not work on TS1 for me aswell, no iframe is displayed and in the moodle course with iframes it does not work

As far as I can see this issue is still not solved. To what setting is the "Default launch container" option in the LTI configuration of the moodle test course "TS1 - Artemis Feature Demo Course" set?

**Review by Jan-Thurner — CHANGES_REQUESTED** [review]
This PR still has the same errors as the last time I reviewed it.

https://github.com/ls1intum/Artemis/pull/9231#pullrequestreview-2255529998
https://github.com/ls1intum/Artemis/pull/9231#issuecomment-2308440068

**Review by pzdr7 — COMMENTED** [review]
Tested on TS1. Clicking on an exercise in Moodle forwards me to the correct exercise on the test server, but this doesn't sound like the intended behavior ("opened in an iFrame") to me. Is that the correct behavior?

**Comment by raffifasaro** [comment]
iFrame issue should be fixed now, pr can be tested and reviewed again

File: src/main/java/de/tum/in/www1/artemis/service/connectors/lti/LtiService.java [hunk 1]
```diff
@@ -105,6 +109,15 @@ public void authenticateLtiUser(String email, String username, String firstName,
         // 2. Case: Lookup user with the LTI email address and make sure it's not in use
         if (artemisAuthenticationProvider.getUsernameForEmail(email).isPresent() || userRepository.findOneByEmailIgnoreCase(email).isPresent()) {
             log.info("User with email {} already exists. Email is already in use.", email);
+
+            if (trustExternalLTISystems) {
+                // If the email is already in use, but we trust external LTI systems, we authenticate the user with the email
```
Comment (az108) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/lti/LtiService.java:114]: This comment is kind of unnecessary as we have a if and a log info before this line to describe the same things as the comment

File: src/main/java/de/tum/in/www1/artemis/service/connectors/lti/LtiService.java [hunk 2]
```diff
@@ -179,23 +192,16 @@ private void addUserToExerciseGroup(User user, Course course) {
...
+            log.info("User is not activated. Adding initialize parameter to query.");
             uriComponentsBuilder.queryParam("initialize", "");
         }
-        else {
-            log.info("User is activated. Adding JWT cookie for logout.");
-            prepareLogoutCookie(response);
-            uriComponentsBuilder.queryParam("ltiSuccessLoginRequired", user.getLogin());
-        }
+
+        log.info("Add/Update JWT cookie so the user will be logged in");
```
Comment (az108) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/lti/LtiService.java:202]: To keep it consistent we should add a fullstop at the end here.
`log.info("Add/Update JWT cookie so the user will be logged in.");`

**Review by az108 — CHANGES_REQUESTED** [review]
Code Looks good overall, just 2 small comments. Tested on TS1 and works as expected 👍 
image

**Review by Jan-Thurner — COMMENTED** [review]
Tested on TS1 as TestUser2. Now it works as described. 
image
Gode LGTM. Thanks for implementing the requested Changes.

**Review by Jan-Thurner — APPROVED** [review]
Approve to previous comment.

**Review by pzdr7 — APPROVED** [review]
Authentication & iFrame work as described on TS1 👍

**Review by az108 — APPROVED** [review]
Thanks for the quick changes. Code LGTM now 👍

**Review by raffifasaro — APPROVED** [review]
Tested on TS1, works as described.

**Review by MaximilianAnzinger — APPROVED** [review]
Maintainer approved

**Pull request merged**